[MRG+1] Add an example and a method to analyse the decision tree stucture

arjoly opened the pull request:
Ping @glouppe, @pprett, @ogrisel, @amueller 

Suggestions are welcome to improve the example. 

It should fix #1105 and #5441.

arjoly committed:
Implement decision path in tree

arjoly committed:
DOC add an example highlighting the tree structure

glouppe commented on examples/tree/plot_structure.py:
```diff
@@ -0,0 +1,110 @@
+
+# First let's retrieve the decision path of each sample. The decision_paths
+# method allows to retrieve the node indicator function. A non zero elements at
+# position (i, j) indicates that the sample i goes # through the node j.
```
On Python 2, this yields an error.

```
➜  scikit-learn git:(9c2af10) python examples/tree/plot_structure.py 
  File "examples/tree/plot_structure.py", line 66
SyntaxError: Non-ASCII character '\xc2' in file examples/tree/plot_structure.py on line 66, but no encoding declared; see http://python.org/dev/peps/pep-0263/ for details
```

(Everything works fine under Python 3 though)

glouppe commented:
It would be nice to include in the example a visualization of the tree, as done with `export_graphviz`.

glouppe commented on examples/tree/plot_structure.py:
```diff
@@ -0,0 +1,110 @@
+Understanding the decision tree structure
+=========================================
+
+The decision tree structure could be analysed to gain further insight on the
```
could -> can

glouppe commented on examples/tree/plot_structure.py:
```diff
@@ -0,0 +1,110 @@
+relation between the features and the target to predict. In this example, we
+show how to retrieve:
+    - the binary tree structure;
+    - the nodes that were reaches by a sample using the decision_paths method;
```
reached

glouppe commented on examples/tree/plot_structure.py:
```diff
@@ -0,0 +1,110 @@
+show how to retrieve:
+    - the binary tree structure;
+    - the nodes that were reaches by a sample using the decision_paths method;
+    - the leaf that was reaches by a sample using the apply method;
```
reached

glouppe commented on examples/tree/plot_structure.py:
```diff
@@ -0,0 +1,110 @@
+estimator.fit(X_train, y_train)
+
+# The decision estimator has an attribute called tree_  which stores the entire
+# tree structure and allow to access to low level attribute. The binary tree
```
allows

glouppe commented on examples/tree/plot_structure.py:
```diff
@@ -0,0 +1,110 @@
+#   - threshold, threshold value at the node
+#
+
+# Using those array, we can parse the tree structure:
```
arrays

glouppe commented on examples/tree/plot_structure.py:
```diff
@@ -0,0 +1,110 @@
+      "the following tree structure:"
+      % estimator.tree_.node_count)
+
+for i in np.arange(estimator.tree_.node_count):
```
is there a reason not to use `range` simply?

glouppe commented on examples/tree/plot_structure.py:
```diff
@@ -0,0 +1,110 @@
+             estimator.tree_.feature[node_id],
+             X_test[i, estimator.tree_.feature[node_id]],
+             threshold_sign,
+             estimator.tree_.threshold[node_id]))
```
I would print `node_id` as well, to make it easier to follow where we are in the tree (which has been printed before).

arjoly replied:
Done

jmschrei replied:
Is it possible to include tabs mimicking the tree structure, or do you think that might be confusing for very large trees?

arjoly replied:
It's possible however I will need to have the depth of each node which is not trivial.

glouppe commented on sklearn/tree/tree.py:
```diff
@@ -454,6 +454,30 @@ def apply(self, X, check_input=True):
+        -------
+        indicator : sparse csr array, shape = [n_samples, n_nodes]
+            Return a node indicator matrix where non zero elements
+            indicates that the samples goes through the samples.
```
through the nodes*?

arjoly committed:
take glouppe comment

arjoly commented:
@glouppe I have taken your comments into account.

arjoly commented:
ping @jmschrei

arjoly commented:
ping @jnothman

glouppe commented:
> It would be nice to include in the example a visualization of the tree, as done with export_graphviz.

What about this? I am not sure it easily feasible though, since the example gallery expect plots to be generated from matplotlib... (as far as I know)

glouppe commented:
Other than that, the rest of the code looks good to me. Thanks for this!

jmschrei commented on sklearn/tree/_tree.pyx:
```diff
@@ -867,6 +865,166 @@ cdef class Tree:
 
         return out
 
+    cpdef object decision_paths(self, object X):
```
Might want to mention that it returns a sparse matrix, and the reason (because it has to have as many columns as the maximal path)

jmschrei replied:
Also, I think it should be `decision_path`, not `decision_paths`

jmschrei commented on sklearn/tree/tests/test_tree.py:
```diff
@@ -1325,3 +1335,36 @@ def test_presort_sparse():
+
+
+def check_decision_path(name):
+    X = iris.data
```
Is it possible to include a small example with a hardcoded decision path? That'll ensure we're getting the right result. Maybe just a training set with 4-5 points, and a testing set of 2 points.

arjoly replied:
Good idea!

jmschrei commented on sklearn/tree/tree.py:
```diff
@@ -454,6 +454,30 @@ def apply(self, X, check_input=True):
         X = self._validate_X_predict(X, check_input)
         return self.tree_.apply(X)
 
+    def decision_paths(self, X, check_input=True):
```
I think this should be `decision_path`, not `decision_paths`. I don't like the plural. Minor comment, though.

jmschrei commented:
Do you think it might be worthwhile to wrap some of the attributes of _tree in the `DecisionTreeRegressor/Classifier` object? Such as `node_count`, and maybe even the arrays? While it makes sense to me why we have a tree object underlying a tree object, it might be easier for others.

jmschrei commented:
> It would be nice to include in the example a visualization of the tree, as done with export_graphviz.

It would be nice to have an option where you can pass in a tree, and a path, and get a picture of the tree with the path highlighted. I understand this may be out of scope for this PR though.

arjoly commented:
> Do you think it might be worthwhile to wrap some of the attributes of _tree in the DecisionTreeRegressor/Classifier object? Such as node_count, and maybe even the arrays? While it makes sense to me why we have a tree object underlying a tree object, it might be easier for others

I don't have a strong opninion. However, I believe it will complicate the overall code by adding many properties.

arjoly commented:
>> It would be nice to include in the example a visualization of the tree, as done with export_graphviz.

> It would be nice to have an option where you can pass in a tree, and a path, and get a picture of the tree with the path highlighted. I understand this may be out of scope for this PR though

This is a good idea however I don't see how to make this with tools such as matplotlib.


>> It would be nice to include in the example a visualization of the tree, as done with export_graphviz.

> What about this? I am not sure it easily feasible though, since the example gallery expect plots to be generated from matplotlib... (as far as I know)

I forgot this one. I can use export_graphviz, but I fear that I won't get anything better than a list of unreadable strings.

glouppe commented:
> I forgot this one. I can use export_graphviz, but I fear that I won't get anything better than a list of unreadable strings.

We could maybe convert it to an image and then "plot" it using `plt.imshow`?

arjoly committed:
DOC make visualizatoin aware of the tree depth

arjoly committed:
ENH rename decision_paths to decision_path

arjoly committed:
Add an hardcoded test

arjoly committed:
typo

arjoly commented:
@jmschrei I have renamed the function, added `depth * '\t'` when I plot the decision tree in the example and hardcoded some node indicator value.

@glouppe I am not sure that this is possible to use graphviz on the doc builder. Maybe this should be made in another pr.

arjoly commented:
I am thinking of adding a decision path in the forest. This could be useful to generate new feature from the forest.

arjoly committed:
Add a decision_path function to forest estimator

arjoly commented:
this should be ready for a new round of review.

glouppe commented:
> @glouppe I am not sure that this is possible to use graphviz on the doc builder. Maybe this should be made in another pr.

Okay, fair enough. I'll see what I can do once this is merged.

glouppe commented on examples/tree/plot_structure.py:
```diff
@@ -0,0 +1,134 @@
+show how to retrieve:
+    - the binary tree structure;
+    - the depth of each node and whether or not it's a leaf.
+    - the nodes that were reached by a sample using the decision_paths method;
```
decision_path, without s

glouppe commented on examples/tree/plot_structure.py:
```diff
@@ -0,0 +1,134 @@
+                 ))
+print()
+
+# First let's retrieve the decision path of each sample. The decision_paths
```
path

glouppe commented on examples/tree/plot_structure.py:
```diff
@@ -0,0 +1,134 @@
+    if is_leaves[node_id]:
+        continue
+
+    if (X_test[i, feature[node_id]] <= threshold[node_id]):
```
it should be `X_test[sample_id, ...`, shouldnt it?

glouppe commented:
+1 for merge once my last comments are fixed.

arjoly committed:
Take into account glouppe comments

arjoly commented on sklearn/ensemble/forest.py:
```diff
@@ -182,6 +184,39 @@ def apply(self, X):
+        n_nodes.extend([i.shape[1] for i in indicators])
+        n_nodes_ptr = np.array(n_nodes).cumsum()
+
+        return sparse_hstack(indicators).tocsr(), n_nodes_ptr
```
@glouppe Should I keep the original format coo or make the conversion to csr? In some case, we might want a csc matrix.

glouppe replied:
I am happy with csr, as for single trees.

jmschrei replied:
In what cases would we want a csc matrix? Given that you can still convert from csr to csc, I think csr is the preferable option here.

jmschrei replied:
Can you also remind me what the purpose of n_nodes_ptr is? Is this information not stored in the csr matrix?

arjoly replied:
`n_nodes_ptr` gives the slice of the csr matrix where you can have the node indicator related to a given __estimator__,
while the indprt from the csr matrix tells you the slice where you have the non zero elements for each __samples__.

arjoly commented:
Thanks @glouppe for the review!

amueller commented on examples/tree/plot_structure.py:
```diff
@@ -0,0 +1,135 @@
+
+The decision tree structure can be analysed to gain further insight on the
+relation between the features and the target to predict. In this example, we
+show how to retrieve:
```
not sure how this renders on the website. Have you checked?

arjoly replied:
Now it renders correctly.

amueller commented on examples/tree/plot_structure.py:
```diff
@@ -0,0 +1,135 @@
+relation between the features and the target to predict. In this example, we
+show how to retrieve:
+    - the binary tree structure;
+    - the depth of each node and whether or not it's a leaf.
```
s/./; ?

amueller commented on examples/tree/plot_structure.py:
```diff
@@ -0,0 +1,135 @@
+
+common_node_id = np.arange(n_nodes)[common_nodes]
+
+print("\nThe following sample %s shares the following node %s  in the tree"
```
Should it be "the samples %s share the node %s in the tree"?

amueller commented on examples/tree/plot_structure.py:
```diff
@@ -0,0 +1,135 @@
+    else:
+        threshold_sign = ">"
+
+    print("rule %s from node %s : (X[%s, %s] (= %s) %s %s)"
```
Maybe I wouldn't call it "rule %s from node %s". This sounds like there is multiple rules in each node. Maybe "decision X in node X"?

amueller commented on examples/tree/plot_structure.py:
```diff
@@ -0,0 +1,135 @@
+
+print("\nThe following sample %s shares the following node %s  in the tree"
+      % (sample_ids, common_node_id))
+print("It is %s %% of all nodes."
```
float division ;)

arjoly committed:
Take into account amueller comments

arjoly committed:
wording

arjoly commented:
I have taken into account your comment @amueller.

amueller commented:
one of the three ;)

arjoly committed:
Rename since we don't plot anything

arjoly commented:
Now it's good;

amueller commented:
thanks :) I didn't review the ``decision_path`` (yet?) so I don't want to give a +1.

ngoix commented on examples/tree/unveil_tree_structure.py:
```diff
@@ -0,0 +1,133 @@
+
+# First let's retrieve the decision path of each sample. The decision_path
+# method allows to retrieve the node indicator function. A non zero elements at
+# position (i, j) indicates that the sample i goes sthrough the node j.
```
elements -> element
sthrough -> through

ngoix commented on examples/tree/unveil_tree_structure.py:
```diff
@@ -0,0 +1,133 @@
+
+# Now, it's possible to get the tests that were used to predict a sample or
+# a group of samples. First, let's make it for the  sample.
+
```
2 whitespaces before 'sample'

ngoix commented:
I find the example very useful !

jmschrei commented on sklearn/tree/tests/test_tree.py:
```diff
@@ -1325,3 +1335,44 @@ def test_presort_sparse():
+    y = iris.target
+    est = DecisionTreeClassifier(random_state=0, max_depth=1).fit(X, y)
+    node_indicator = est.decision_path(X[:2]).toarray()
+    assert_array_equal(node_indicator, [[1, 1, 0], [1, 0, 1]])
```
I think I am misunderstanding these paths. How can the first point go to node 1, then node 1 again, then node 0? My understanding was that the decision path was an array of node IDs, ending in a leaf node.

glouppe replied:
It is the binary encoded version of this. If `path[i] == 1`, then the sample traverses node with node_id `i`

jmschrei replied:
Okay. So then the later nodes always correspond to later nodes visited, so using it as a mask on the nodes gives you the path. Got it.

arjoly committed:
Take into account ngoix comment

glouppe commented:
Anybody else for a review? or @jmschrei @ngoix can give yours +1 if you think this is ready?

(It would be nice to have this merged quickly, so that work on #4163 can continue.)

ngoix commented on sklearn/ensemble/tests/test_forest.py:
```diff
@@ -1098,3 +1098,32 @@ def test_dtype_convert(n_classes=15):
+        leave_indicator = [indicator[i, n_nodes_ptr[est_id] + j]
+                           for i, j in enumerate(leaves[:, est_id])]
+        assert_array_almost_equal(leave_indicator, np.ones(shape=n_samples))
+
```
I don't understand the almost_equal

ngoix commented:
For what I understood it seems good to me but I'm not fluent enough in cython to give a +1.

agramfort commented:
Thanks
